src/factories/nodes.ts

```diff
@@ -152,7 +167,55 @@ export async function nodesContainerFactory() {
     edgesContainer.addChild(edge.element)
   }
 
-  function getLayout(data: RunGraphData): void {
+  async function createArtifacts(): Promise<void> {
+    const { artifacts } = dataBundle ?? {}
+
+    if (!artifacts) {
+      return
+    }
+
+    const promises: Promise<Container>[] = []
+    const artifactGroups = groupArtifactsByNodeId(artifacts)
+
+    for (const [nodeId, nodeArtifacts] of artifactGroups) {
+      const node = nodes.get(nodeId)
+
+      if (!node) {
+        console.warn(`Could not find node for artifacts: Skipping ${nodeId}`)
+        continue
+      }
+
+      promises.push(node.renderArtifacts(nodeArtifacts))
+    }
+
+    await Promise.all(promises)
+  }
+
+  function groupArtifactsByNodeId(artifacts: Artifact[]): Map<string, Artifact[]> {
+    const grouped = new Map<string, Artifact[]>()
+
+    for (const artifact of artifacts) {
+      if (!artifact.taskRunId) {
+        continue
+      }
+
+      const group = grouped.get(artifact.taskRunId) ?? []
+
+      group.push(artifact)
+
+      grouped.set(artifact.taskRunId, group)
+    }
+
+    return grouped
+  }
```

Comment:
This grouping logic could be abstracted into a utility in a separate file. That way we can keep this file smaller.